### ISSUE TITLE
Cannot read property 'remoteAddress' of undefined

### ISSUE THREAD

**Issue by gregorwolf** [sapmentors/cap-community#31]
Hello CAP Enthusiasts,

during local testing I'm facing the following issue:

In the test project [gregorwolf/bookshop-demo](https://github.com/gregorwolf/bookshop-demo) I've defined two users in [.cdsrc.json](https://github.com/gregorwolf/bookshop-demo/blob/master/.cdsrc.json): admin and littleadmin. Only admin has the scope roleadmin which is used to restrict the access to the entity Roles of the [admin-service.cds](https://github.com/gregorwolf/bookshop-demo/blob/master/srv/admin-service.cds). 

To replicate the issue please do the following:

1. Start the app with `npm run build:deploy:startv2`
2. Open http://localhost:4004/app/fiori.html in a browser
3. Click on the tile Approvals in the Group OData V4
4. You should be prompted for a username and passwort
5. Enter littleadmin as the username. The pasword can be left empty
6. You should get a list of items to be approved
7. Now navigate back to the homepage
8. Click on the tile Role in the Group OData V4
9. Now the backend will crash with the following error message:

```
[2020-01-19T08:10:27.442Z | ERROR | 1427824]: Cannot read property 'remoteAddress' of undefined
/Users/gwolf/Documents/Projects/cap/bookshop-demo/node_modules/@sap/cds-services/lib/adapter/odata-v4/handlers/error.js:41
        throw err
        ^

TypeError: Cannot read property 'remoteAddress' of undefined
    at getIpFromRequest (/Users/gwolf/Documents/Projects/cap/bookshop-demo/node_modules/@sap/cds-services/lib/services/utils/clientFromRequest.js:35:25)
```

The root cause is that in the getIpFromRequest method the variable req.connection.remoteAddress isn't checked for undefined.

Looking forward for a fix.

Best regards
Gregor

**Comment by gregorwolf** [comment]
This issue is now also reported via the SAP Service Marketplace: 40436 / 2020

**Comment by David-Kunz** [comment]
Hi Gregor,

Thanks for reporting, we already fixed that!

The fix will be shipped in the next release.

Best regards,
David

**Comment by gregorwolf** [comment]
Unfortunately not in the just released 3.21.2.

**Comment by David-Kunz** [comment]
Hi Gregor,

Unfortunately, no, but in version 1.23.0 the fix will be included. Hang in there :)

Best regards,
David

**Comment by gregorwolf** [comment]
Hi David,

I do face this issue also in production if a user that isn't assigned to the required scope calls a service that is protected. That causes the srv app to crash. Do you have a clue when 1.23.0 of @sap/cds-services will be released?

Best regards
Gregor

**Comment by David-Kunz** [comment]
Hi Gregor,

Unfortunately, for this new release it took as quite a long time, however we are working hard to release very soon!

Best regards,
David

**Comment by gregorwolf** [comment]
Seems to be fixed with @sap/cds 3.31.2
